admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/SecurityConfig.kt

```diff
@@ -54,25 +70,119 @@ class SecurityConfig(override val di: DI) : DIAware {
+                    name = "keycloak",
+                    authorizeUrl = "$keycloakAddress/realms/master/protocol/openid-connect/auth",
+                    accessTokenUrl = "$keycloakAddress/realms/master/protocol/openid-connect/token",
+                    requestMethod = HttpMethod.Post,
+                    clientId = "drill4j",
+                    clientSecret = "u8pI2hH3dZhsSL1K6kzGsJRQMxj5EcNn",
+                    defaultScopes = listOf("roles", "openid")
+                )
+            }
+            client = HttpClient(Apache)
+        }
     }
 }
 
+fun Routing.oauthRoutes() {
+    authenticate("oauth") {
+        get("/oauth/login") {
+            // Redirects to "authorizeUrl" if authentication fails
+            // or redirect to root otherwise
+            call.respondRedirect("/")
+        }
+        get("/oauth/logout") {
+            call.principal<OAuthAccessTokenResponse.OAuth2>()?.let { principal ->
+                val httpClient = HttpClient(Apache)
+                httpClient.post<HttpResponse>("http://localhost:8080/realms/master/protocol/openid-connect/logout") {
+                    headers {
+                        append(HttpHeaders.Authorization, "Bearer ${principal.accessToken}")
+                    }
+                }
+            }
+        }
+        get("/oauth/callback") {
+            val principal: OAuthAccessTokenResponse.OAuth2? = call.principal()
+            if (principal != null) {
+                getUserInfo(principal.accessToken)
+
+                call.response.cookies.append(Cookie("jwt", principal.accessToken, httpOnly = true, path = "/"))
+                call.respondRedirect("/")
+            } else {
+                call.unauthorizedError()
+            }
+        }
+    }
+}
+
+private suspend fun getUserInfo(accessToken: String): User? {
+    val httpClient = HttpClient(Apache)
+    val response =
+        httpClient.get<HttpResponse>("http://localhost:8080/realms/master/protocol/openid-connect/userinfo") {
+            headers {
+                append(Authorization, "Bearer $accessToken")
+            }
+        }
+    return response.status.takeIf { it.isSuccess() }?.let { _ ->
+        val stringBody = response.receive<String>()
+        val jsonBody = Json.parseToJsonElement(stringBody)
+        jsonBody.toPrincipal()
+    }
+}
+
+private fun JsonElement.toPrincipal(): User {
+    val jsonObject = this.jsonObject
+    return User(
+        username = jsonObject["username"]?.jsonPrimitive?.contentOrNull ?: "",
+        role = findRole(jsonObject["roles"]?.jsonArray?.map { it.toString() } ?: emptyList())
+    )
+}
+
 private fun Payload.toPrincipal(): User {
     return User(
         username = subject,
         role = Role.valueOf(getClaim("role").asString())
     )
 }
 
+private fun findRole(roleNames: List<String>): Role {
+    return Role.values().find { role ->
+        roleNames.toSet().map { it.lowercase() }.contains(role.name.lowercase())
+    } ?: Role.UNDEFINED
+}
```

Comment:
is this some sort of the "mapping", that matches the first match from "external" roles to "our" roles?

Reply:
Yes. But this mapping will be improved in the non-POC version